File: leetcode-jvm/src/main/java/com/lone/leetcode/j/Solution8.java
package com.lone.leetcode.j;

/**
 * https://leetcode.com/problems/string-to-integer-atoi/
 * @author lone
 * @date 2018/10/9
 */
public class Solution8 {

    public int myAtoi(String str) {
        int index = 0;
        int result = 0;
        boolean positive = true;
        while (index < str.length()) {
            char c = str.charAt(index);
            if (c == ' ') {
                index++;
                continue;
            } else if (c == '+') {
                index++;
                break;
            } else if (c == '-') {
                positive = false;
                index++;
                break;
            } else {
                break;
            }
        }
        if (index >= str.length()) {
            return result;
        }
        while (index < str.length()) {
            int tmp = str.charAt(index) - '0';
            if (tmp >= 0 && tmp <= 9) {
                tmp = (positive? 1 : -1) * tmp;
                if (isValid(result, tmp, positive)) {
                    result = result * 10 + tmp;
                    index++;
                } else if (positive) {
                    result = Integer.MAX_VALUE;
                    break;
                } else {
                    result = Integer.MIN_VALUE;
                    break;
                }
            } else {
                break;
            }
        }
        return result;
    }

    private boolean isValid(int result, int tmp, boolean positive) {
        if(positive) {
            return result <= Integer.MAX_VALUE / 10 && (tmp < (Integer.MAX_VALUE - result * 10));
        } else {
            return result >= Integer.MIN_VALUE / 10 && (tmp > Integer.MIN_VALUE - result * 10);
        }
    }

    public static void main(String[] args) {
        Solution8 s = new Solution8();
        System.out.println(s.myAtoi("3.14159"));
    }
}

File: leetcode-jvm/src/main/kotlin/com/lone/leetcode/k/Solution8.kt
package com.lone.leetcode.k

/**
 * https://leetcode.com/problems/string-to-integer-atoi/
 * @author lone
 * @date  2018/10/8
 */
class Solution8 {
    fun myAtoi(str: String): Int {
        var ret = 0
        var positive = true
        var start = 0
        while (start < str.length) {
            val cur = str[start]
            if (cur == ' ') {
                start++
                continue
            } else if (cur == '-') {
                positive = false
                start++
                break
            } else if (cur == '+') {
                start++
                break
            } else {
                break
            }
        }
        if (start >= str.length) {
            return 0
        }
        while (start < str.length) {
            val cur = str[start++] - '0'
            if (cur !in 0..9) {
                break
            }
            val i = if(positive) cur  else cur * -1
            if(isValid(ret,i,positive))  {
                ret = ret * 10 + i
            }else if (positive) {
                ret = Int.MAX_VALUE
                break
            } else {
                ret = Int.MIN_VALUE
                break
            }
        }
        return ret
    }

    private fun isValid(ret: Int, addon: Int,positive: Boolean):Boolean {
        return if (positive) {
            ret  <= Int.MAX_VALUE/10 && addon <= (Int.MAX_VALUE - ret * 10)
        } else {
            ret  >= Int.MIN_VALUE/10 && addon >= (Int.MIN_VALUE - ret * 10)
        }
    }
}

fun main(args: Array<String>) {
 val so8 = Solution8()
    println(so8.myAtoi("2147483646"))
//    -538412641*10
}
